### ISSUE TITLE
cross-attention in llama 3.2 vision has is_causal=False

### ISSUE THREAD

**Issue by mmehdig** [pytorch/torchtune#2493]
The definition of the attention module for cross-attention between vision and text explicitly specifies `is_causal=False` I wonder why?

Here: https://github.com/pytorch/torchtune/blob/d5d12fef1f8c39dfd5c9f85807795ef503216e12/torchtune/models/llama3_2_vision/_component_builders.py#L212-L227

Shouldn't it be the opposite? 

The cross-attention module is part of the decoder (the token prediction), so based on my understanding, it should not be able to see the future text. Q refers to the target text, and KV is the image source in this case. So, we either need to set `is_causal=True` or the cross-attention mask must handle the masking of the future text for fine-tuning. I don't see any of these two happening.

I may be missing some aspects of the preprocessing. My resources are limited for running proper tests here can someone else also have a look?

**Comment by RdoubleA** [comment]
The cross attention between text and vision is not text tokens attending to past or future text tokens. If that's ere the case then you would be correct, it should be causal. Instead it is text attending to vision tokens, which doesn't follow the causal pattern. This requires a separate cross attention mask to be passed in that determines which images tokens are attended to by each text token. Since we are using a custom attention mask, is_causal must be false.

You can see [the docs here](https://pytorch.org/torchtune/main/generated/torchtune.modules.transforms.VisionCrossAttentionMask.html#torchtune.modules.transforms.VisionCrossAttentionMask) on the VisionCrossAttentionMask for how it would look like.

Let me know if things still aren't clear.

**Comment by mmehdig** [comment]
> The cross attention between text and vision is not text tokens attending to past or future text tokens. If that were the case then you would be correct, it should be causal. Instead it is text attending to vision tokens, which doesn't follow the causal pattern. This requires a separate cross attention mask to be passed in that determines which image tokens are attended to by each text token. Since we are using a custom attention mask, is_causal must be false.

This is precisely related to my claim. The custom attention mask we are generating [here](https://github.com/pytorch/torchtune/blob/d5d12fef1f8c39dfd5c9f85807795ef503216e12/torchtune/models/llama3_2_vision/_transform.py#L216) can see the future text at the point of fusion. Because the samples at training time have all the future tokens, then our custom cross-attention mask is the same for all tokens.

**Comment by mmehdig** [comment]
To be clear, based on @RdoubleA's answer, I now understand why we need to set causal=False. but still, it doesn't answer the fundamental question about accessing future tokens from my point of view. 
The role of cross-attention is to contextualize the encoded vision embeddings based on the text tokens, then it is fused with text tokens. however, my claim is that this attention module seems to be able to contextualize vision embeddings while accessing the future context during the training time. Potentially, this is not a big deal and it is hard to see any effect when we do small supervise fine-tuning jobs. 

To support my claim, this is an explanation in stack exchange about cross-attention:
https://datascience.stackexchange.com/a/126213

**Comment by mmehdig** [comment]
> To support my claim, this is an explanation in stack exchange about cross-attention: https://datascience.stackexchange.com/a/126213

Now I understand why this is the case. The causal mask that hides the future tokens, in fact, is about the **source sequence** (kv sequence dimension), not the target sequence. The dot product `query @ keyT` explains it. The query doesn't need masking because they never cross paths. 

The vision input is timed with this implementation based on its appearance in text, so things look fine 👍

**Comment by RdoubleA** [comment]
@mmehdig Sorry I never got around to following up on this! Glad things make sense now, please let me know if there's anything else I can help clarify

**Comment by mmehdig** [comment]
@RdoubleA thanks :)
Yes I was about to post another issue. It seems we actually don't use this mask it is disabled here:
https://github.com/pytorch/torchtune/blob/3fc4c95bef9911da907d9403c9196c024a42089f/torchtune/modules/transformer.py#L284
I believe this commit changed `0` to `True`. I have a feeling that it was intentional. but the consequence is that the mask is all True on text. it cancels the mask generated here.

**Comment by RdoubleA** [comment]
Yes if you don't mind let's post this as another separate issue so it gets more eyes
